plans/smlbench2/main.go

```diff
@@ -0,0 +1,232 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"time"
+
+	utils "github.com/ipfs/testground/plans/smlbench2/utils"
+	iptb "github.com/ipfs/testground/sdk/iptb"
+	"github.com/ipfs/testground/sdk/runtime"
+	"github.com/ipfs/testground/sdk/sync"
+	"github.com/libp2p/go-libp2p-core/peer"
+	// shell "github.com/ipfs/go-ipfs-api"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/ipfs/go-cid"
+)
+
+const (
+	// sizeBytes = 10 * 1024 * 1024 // 10mb
+	sizeBytes = 100 * 1024 * 1024 // 100mb
+)
+
+var cidSubtree = &sync.Subtree{
+        GroupKey:    "cid",
+        PayloadType: reflect.TypeOf(&cid.Cid{}),
+        KeyFunc: func(val interface{}) string {
+                return "cid"
+        },
+}
+
+func main() {
+	runenv := runtime.CurrentRunEnv()
+	if runenv.TestCaseSeq < 0 {
+		panic("test case sequence number not set")
+	}
+
+	timeout := func() time.Duration {
+                if t, ok := runenv.IntParam("timeout_secs"); !ok {
+                        return 30 * time.Second
+                } else {
+                        return time.Duration(t) * time.Second
+                }
+        }()
+
+	watcher, writer := sync.MustWatcherWriter(runenv)
+        defer watcher.Close()
+
+	spec := iptb.NewTestEnsembleSpec()
+	spec.AddNodesDefaultConfig(iptb.NodeOpts{Initialize: true, Start: true}, "local")
+
+	ctx := context.Background()
+	ensemble := iptb.NewTestEnsemble(ctx, spec)
+	ensemble.Initialize()
+
+	localNode := ensemble.GetNode("local")
+
+	peerID, err := localNode.PeerID()
+	if err != nil {
+		runenv.Abort(err)
+	}
+	fmt.Println("Jim peerID", peerID)
+
+	swarmAddrs, err := localNode.SwarmAddrs()
+	if err != nil {
+		runenv.Abort(err)
+	}
+	fmt.Println("Jim swarmAddrs", swarmAddrs)
+
+	ID, err := peer.IDB58Decode(peerID)
+	if err != nil {
+		runenv.Abort(err)
+	}
+
+	addrs := make([]ma.Multiaddr, len(swarmAddrs))
+	for i, addr := range swarmAddrs {
+		multiAddr, err := ma.NewMultiaddr(addr)
+		if err != nil {
+			runenv.Abort(err)
+		}
+		addrs[i] = multiAddr
+	}
+	addrInfo := &peer.AddrInfo{ID, addrs}
+
+	seq, err := writer.Write(sync.PeerSubtree, addrInfo)
+	if err != nil {
+                runenv.Abort(err)
+        }
+        defer writer.Close()
+
+	client := localNode.Client()
+
+	// States
+	added := sync.State("added")
+	received := sync.State("received")
+
+	switch {
+	case seq == 1: // adder
+		adder := client
+		fmt.Println("adder")
+
+		// generate a random file of the designated size.
+		file := utils.TempRandFile(runenv, ensemble.TempDir(), sizeBytes)
+		defer os.Remove(file.Name())
+
+		tstarted := time.Now()
+		hash, err := adder.Add(file)
+		if err != nil {
+			runenv.Abort(err)
+			return
+		}
+		fmt.Println("cid", hash)
+
+		c, err := cid.Parse(hash)
+		if err != nil {
+			runenv.Abort(err)
+			return
+		}
+
+		runenv.EmitMetric(utils.MetricTimeToAdd, float64(time.Now().Sub(tstarted)/time.Millisecond))
```

Comment:
Note that for this Plan. The time adding the file isn't important. Only the time transferring it. (+ the other things to measure, such as number of repeated blocks, etc as described in the spec for the plan)

Reply:
I just copied that from the original smlbench test. I'm not even really sure yet what schema for the JSON is going to work well with graphing things in Kibana - I think we need to try to build some reports/graphing there really soon.

I'd love to dump as many metrics, events, traces as we possibly can, but tag things in an understandable way so we can easily generate reports with the most important metrics that we care about. It should be clear in the test what the primary metrics are vs. secondary metrics.